### DISCUSSION TITLE
Using data from one useContractRead as argument to another

### DISCUSSION THREAD

**Discussion by TedDDDao** [wevm/wagmi#1121] [Q&A] [answered]
I'm finding useContractRead very difficult to use compared to simply calling the contract function directly, because I can't easily control the timing of when the function gets called, or what the pre-call arguments are.

I currently need to call two contract functions in the same component, let's call them getFoo(), and getBar(foo)

What is the "right" way to do this?  I've tried defining two useContractRead, but I get an error because getBar requires an argument, which it doesn't have when the hook initially runs.

I really don't want to start nesting components just to pass the result from one into the next, as that seems messy.  Should I go to useContract and call the function directly that way?  That seems to negate many of the benefits of wagmi over something like web3-react.

**Comment by tmm** [comment, marked as answer]
You can set up `getBar` to only run after `getFoo` resolves with `useContractRead`'s [`enabled`](https://wagmi.sh/docs/hooks/useContractRead#enabled-optional) option.

```ts
const { data: foo } = useContractRead({
  address: '0x…',
  abi: […],
  functionName: 'getFoo',
})
const { data: bar } = useContractRead({
  address: '0x…',
  abi: […],
  functionName: 'getBar',
  args: foo ? [foo] : undefined,
  enabled: !!foo,
})
```

**Reply by TedDDDao** [reply]
Thank you for the quick reply, this worked for me.

For what it's worth, calling the contract via react hooks seems very restrictive compared to async functions, but I'm trying to keep an open mind.

**Reply by tmm** [reply]
Totally! It's a trade-off: React Hooks are declarative and hook into the lifecycle (`status` changes, request deduplication, etc.), while using core actions (e.g. `import { readContract } from 'wagmi/actions'`) are imperative and you need to manage their state.
